### PR TITLE
Checkout: Only show privacy upsell if all domains support privacy

### PR DESCRIPTION
**Pull request by yoavf**
This is the path of least resistance to handle the fact that some upcoming TLDs do not offer private registrations.

If any of the domains in the cart doesn't support private registrations, the privacy section and privacy dialogue won't be shown at all.

In the future, we can work out the UI and logic for adding privacy to the domains that support it.

- [x] TODO: add tests

### CODE REVIEW

File: client/my-sites/upgrades/checkout/domain-details-form.jsx [hunk 1]
```diff
@@ -399,7 +403,7 @@ class DomainDetailsForm extends Component {
 	}
 
 	finish( options = {} ) {
-		this.setPrivacyProtectionSubscriptions( options.addPrivacy !== false );
+		this.setPrivacyProtectionSubscriptions( options.addPrivacy && options.addPrivacy !== false );
```
Comment (deBhal) [inline client/my-sites/upgrades/checkout/domain-details-form.jsx]: If `options.addPrivacy` is truthy, it can never be `false`, right?

The old code here seems weird - if options.addPrivacy is `null` or `undefined` I wouldn't think we'd want to set the subscriptions. Was it supposed to be a string check or something?

**Comment by deBhal** [comment]
I was hoping to be able to break off at least a little bit of cart and reduxify it, but it's so big and complicated that I had to give it up :(

Instead, I've rebased and added tests for showing the privacy panel.

I also updated the logic in `finish()`. I'm pretty sure it's right now, but I want to talk through my reasoning for the validation:

We skip the dialog and go straight from `handleSubmitButtonClick()` into `this.finish()` with no options in 2 cases:

1)  The user already has privacy in their basket for all their domains
2) The user has a domain that doesn't support privacy.

In both cases, we want to not change the privacy. Turning it off or on would annoy the user or annoy us respectively :) 

So if we come through the other path with the `addPrivacy` from the `privacy-protection-dialog`,  
we're going to call `setPrivacyProtectionSubscriptions()` in both branches, so I've unified that call before the if.

Tl;DR: Looks good to me now. @klimeryk?

**Comment by yoavf** [comment]
@deBhal Thanks for the changes and for adding the tests.  LGTM 👍

File: client/lib/cart-values/cart-items.js [hunk 2]
```diff
@@ -283,6 +283,12 @@ function hasDomainRegistration( cart ) {
 	return some( getAll( cart ), isDomainRegistration );
 }
 
+function hasOnlyDomainRegistrationsWithPrivacySupport( cart ) {
+	return every( getDomainRegistrations( cart ), function( cartItem ) {
```
Comment (klimeryk) [inline client/lib/cart-values/cart-items.js]: Could probably be written as `every( getDomainRegistrations( cart ), privacyAvailable )`?

File: client/lib/cart-values/cart-items.js [hunk 3]
```diff
@@ -404,7 +410,11 @@ function themeItem( themeSlug, source ) {
  * @returns {Object} the new item as `CartItemValue` object
  */
 function domainRegistration( properties ) {
-	return assign( domainItem( properties.productSlug, properties.domain, properties.source ), { is_domain_registration: true } );
+	return assign( domainItem( properties.productSlug, properties.domain, properties.source ),
+		{
+			is_domain_registration: true,
+			...( properties.extra ? { extra: properties.extra } : {} ),
```
Comment (klimeryk) [inline client/lib/cart-values/cart-items.js]: This is actually not needed. Because the shopping cart is magical ;)
OK, seriously - the way I understand it, we create the item on the frontend, send it for validation to the backend. That's were the proper `extra` is written, sometimes additional products are added to the cart if needed, and the new shopping cart is returned. And _that_ one is the "canonical" one. That's how the `registrar` field is used right now.

File: client/lib/cart-values/cart-items.js [hunk 4]
```diff
@@ -714,6 +724,16 @@ function isRenewal( cartItem ) {
 }
 
 /**
+ * Determines whether a cart item supports privacy
+ *
+ * @param {Object} cartItem - `CartItemValue` object
+ * @returns {boolean} true if item supports privacy
+ */
+function privacyAvailable( cartItem ) {
+	return cartItem.extra && cartItem.extra.privacy_available;
```
Comment (klimeryk) [inline client/lib/cart-values/cart-items.js]: I like lodash's `get` for stuff like this - you can easily add a default value too: `return get( cartItem, 'extra.privacy_available', true )`. Yes, that's `true` by default - if it's not declared for some reason, then that means it's an old shopping cart from before this PR and backend changes (though, a trip to the backend should update it). Shopping carts never expire, so it's better to provide some sane defaults ;)

File: client/lib/cart-values/cart-items.js (hunk 3 above)
Comment (yoavf) [inline client/lib/cart-values/cart-items.js]: The tests added here won't pass without it, because the extra field is stripped in domainItem. 
I think this is ok as is, or we can add "native" support for 'extra' to domainItem - whatever you think is better.

Comment (klimeryk) [inline client/lib/cart-values/cart-items.js]: Then that would suggest that the tests are not actually testing how the system behaves in real life :) But I've promised myself not to touch anything that has to do with the shopping cart, so I'm fine with keeping it like this 😜
